File: apps/mobile/modules/sage-capability/ios/SageCapabilityModule.swift
import ExpoModulesCore
import Metal
import UIKit

/**
 * iOS implementation of the SAGE Capability Manifest probe.
 *
 * Platform paradigm (iOS): GPU acceleration via Metal, hardware ML via Core ML
 * / the Apple Neural Engine. Pairs with the Android (Vulkan / NNAPI) module.
 */
public class SageCapabilityModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SageCapability")

    Function("getPlatform") { () -> String in
      "ios"
    }

    Function("getOsVersion") { () -> String in
      UIDevice.current.systemVersion
    }

    // Returned as Double because JS numbers are IEEE-754; RAM fits exactly.
    Function("getTotalRamBytes") { () -> Double in
      Double(ProcessInfo.processInfo.physicalMemory)
    }

    Function("getGpuBackend") { () -> String in
      // llama.cpp uses the Metal backend when a default Metal device exists.
      MTLCreateSystemDefaultDevice() != nil ? "metal" : "none"
    }

    Function("getMlAccelerator") { () -> String in
      // Core ML (and the Apple Neural Engine) ship on every supported device.
      "coreml"
    }

    Function("hasNpu") { () -> Bool in
      // The Apple Neural Engine is present on A11 Bionic and later — i.e. every
      // device that can run iOS 15.1+. Reported true for the deployment target.
      true
    }

    Function("getThermalState") { () -> String in
      switch ProcessInfo.processInfo.thermalState {
      case .critical: return "critical"
      case .serious: return "serious"
      case .fair: return "fair"
      case .nominal: return "nominal"
      @unknown default: return "nominal"
      }
    }

    AsyncFunction("listInstalledModels") { () -> [[String: Any]] in
      SageCapabilityModule.discoverModels()
    }
  }

  private static func discoverModels() -> [[String: Any]] {
    var out: [[String: Any]] = []
    let fm = FileManager.default
    guard
      let docs = fm.urls(for: .documentDirectory, in: .userDomainMask).first
    else { return out }
    let modelsDir = docs.appendingPathComponent("models", isDirectory: true)

    let candidates: [(file: String, id: String)] = [
      ("gemma-4-2b-q4.gguf", "gemma-4-2b"),
      ("gemma-4-9b-q4.gguf", "gemma-4-9b"),
    ]
    for c in candidates {
      let url = modelsDir.appendingPathComponent(c.file)
      guard fm.fileExists(atPath: url.path) else { continue }
      let attrs = try? fm.attributesOfItem(atPath: url.path)
      let size = (attrs?[.size] as? NSNumber)?.intValue ?? 0
      out.append([
        "id": c.id,
        "path": url.path,
        "sizeBytes": size,
        "verified": SageCapabilityModule.verifyGGUF(url: url),
      ])
    }
    return out
  }

  /// A real GGUF file begins with the ASCII magic "GGUF" (0x47 0x47 0x55 0x46).
  private static func verifyGGUF(url: URL) -> Bool {
    guard let handle = try? FileHandle(forReadingFrom: url) else { return false }
    defer { try? handle.close() }
    let magic = try? handle.read(upToCount: 4)
    return magic == Data([0x47, 0x47, 0x55, 0x46])
  }
}

File: apps/mobile/modules/sage-capability/android/src/main/java/com/iterativ/sage/capability/SageCapabilityModule.kt
package com.iterativ.sage.capability

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.os.PowerManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

/**
 * Android implementation of the SAGE Capability Manifest probe.
 *
 * Platform paradigm (Android): GPU acceleration via Vulkan, hardware ML via
 * NNAPI. Pairs with the iOS (Metal / Core ML) module.
 */
class SageCapabilityModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SageCapability")

    Function("getPlatform") { "android" }

    Function("getOsVersion") { Build.VERSION.RELEASE ?: "unknown" }

    Function("getTotalRamBytes") {
      val ctx = appContext.reactContext ?: return@Function 0.0
      val am = ctx.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
      val info = ActivityManager.MemoryInfo()
      am.getMemoryInfo(info)
      info.totalMem.toDouble()
    }

    Function("getGpuBackend") {
      // llama.cpp uses the Vulkan backend on Android; Vulkan 1.1 is broadly
      // available on the API 26+ devices SAGE targets.
      "vulkan"
    }

    Function("getMlAccelerator") {
      // NNAPI is available from API 27 (Android 8.1).
      if (Build.VERSION.SDK_INT >= 27) "nnapi" else "none"
    }

    Function("hasNpu") {
      // Heuristic: SoCs on API 29+ broadly expose an NPU/DSP through NNAPI.
      Build.VERSION.SDK_INT >= 29
    }

    Function("getThermalState") {
      val ctx = appContext.reactContext
      if (ctx != null && Build.VERSION.SDK_INT >= 29) {
        val pm = ctx.getSystemService(Context.POWER_SERVICE) as PowerManager
        when (pm.currentThermalStatus) {
          PowerManager.THERMAL_STATUS_SEVERE,
          PowerManager.THERMAL_STATUS_CRITICAL,
          PowerManager.THERMAL_STATUS_EMERGENCY,
          PowerManager.THERMAL_STATUS_SHUTDOWN -> "critical"
          PowerManager.THERMAL_STATUS_MODERATE -> "serious"
          PowerManager.THERMAL_STATUS_LIGHT -> "fair"
          else -> "nominal"
        }
      } else {
        "nominal"
      }
    }

    AsyncFunction("listInstalledModels") {
      val ctx = appContext.reactContext
        ?: return@AsyncFunction emptyList<Map<String, Any>>()
      val dir = File(ctx.filesDir, "models")
      val candidates = listOf(
        "gemma-4-2b-q4.gguf" to "gemma-4-2b",
        "gemma-4-9b-q4.gguf" to "gemma-4-9b",
      )
      val out = mutableListOf<Map<String, Any>>()
      for ((fileName, id) in candidates) {
        val f = File(dir, fileName)
        if (f.exists()) {
          out.add(
            mapOf(
              "id" to id,
              "path" to f.absolutePath,
              "sizeBytes" to f.length().toDouble(),
              "verified" to verifyGguf(f),
            ),
          )
        }
      }
      out
    }
  }

  /** A real GGUF file begins with the ASCII magic "GGUF" (0x47 0x47 0x55 0x46). */
  private fun verifyGguf(f: File): Boolean {
    return try {
      f.inputStream().use { input ->
        val magic = ByteArray(4)
        if (input.read(magic) != 4) return@use false
        magic[0] == 0x47.toByte() &&
          magic[1] == 0x47.toByte() &&
          magic[2] == 0x55.toByte() &&
          magic[3] == 0x46.toByte()
      }
    } catch (e: Exception) {
      false
    }
  }
}
